rs/replicated_state/src/replicated_state.rs

```diff
@@ -723,17 +738,26 @@ impl ReplicatedState {
             ),
             None => {
                 let subnet_id = self.metadata.own_subnet_id.get_ref();
-                if msg.receiver().get_ref() == subnet_id {
-                    // TODO(MR-523) Assert that this is a request; else check for a matching `Callback`.
-                    push_input(
-                        &mut self.subnet_queues,
-                        msg,
-                        subnet_available_memory,
-                        own_subnet_type,
-                        input_queue_type,
-                    )
+                if receiver.get_ref() == subnet_id {
+                    match &msg {
+                        RequestOrResponse::Request(_) => push_input(
+                            &mut self.subnet_queues,
+                            msg,
+                            subnet_available_memory,
+                            own_subnet_type,
+                            input_queue_type,
+                        ),
+
+                        RequestOrResponse::Response(response) => Err((
+                            StateError::non_matching_response(
+                                "Management canister does not accept canister responses",
+                                response,
+                            ),
+                            msg,
+                        )),
```

Comment:
This is the only functional change in this PR. Returning an error here results in `StreamHandlerImpl` bumping a critical error counter. (Or in `StreamBuilderImpl` panicking, but that would require the management canister to first create a request on its own behalf; and for that request to fail to be routed.)

Reply:
On a second thought: I don't particularly like that this cross component assumption could lead to a crash: execution changing its behavior would trigger a crash in some component that one wouldn't immediately think of. Can we somehow handle it gracefully?

Reply:
> The alternative would be to allow responses in subnet input queues.

Yeah, as said above -- I think we should handle the error case gracefully rather than allowing responses.

> I would rather just make CanisterQueues::push_output_request() pub(super), so it can only be used from within a CanisterState (its only current user is SystemState::push_output_request()); and add a big warning next to it.

Just to make sure I understand correctly: 

- Currently the only way for execution to push something onto the subnet output queues is to use `ReplicatedState::push_subnet_output_response` and there is no corresponding method for requests, meaning that as things are there isn't even a codepath for pushing requests?  
- And making `CanisterQueues::push_output_request` `pub(super)` will add an additional layer of protection from it being used in the public API of replicated state? 

One concern might be that a comment is often not read when making changes, and diffs on PRs are also often collapsing the related lines. But it would certainly be an improvement in terms of safety.

Reply:
I went ahead and reused the critical error counter we had in `StreamHandlerImpl` (by moving it into `MessageRoutingMetrics` and then cloning it into both `StreamBuilderMetrics` and `StreamHandlerMetrics`). It adds more plumbing, but it is what it is...

Reply:
Oh, I also made `CanisterQueues::push_output_request()` `pub(super)` and added the warning, just in case.

To answer your questions, one would have to call `subnet_queues::push_output_request()` in order to enqueue a request on behalf of the management canister. There exists no such code (the only caller of `CanisterQueues::push_output_request()` is `SystemState::push_output_request()`) and now that I've also made it private to the `canister_state` module, a `subnet_queues::push_output_request()` would not work without changing the visibility back to public.

Reply:
Nice. Thanks for the change and for the clarification. I suppose this also means that we don't strictly need to wait for @dsarlis' confirmation. If there is no code path to push requests onto the subnet queue then it is clear that this is what execution's intention is.

Reply:
It's correct that we don't expect responses to be pushed onto subnet input queues. The only case that comes close is the responses delivered by consensus but they are handled in a separate queue (consensus_queue). We do have a ticket to handle them through subnet_queues instead but it's unclear when we'll work on it.